### PR TITLE
Read/write nested dictionary under map in ipc stream reader/writer

### LINKED ISSUE #1582: Read/write nested dictionary under map in ipc stream reader/write
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

#1565 added the support of nested dictionaries under List in IPC stream reader/writer. This is for nested dictionaries under Map in IPC stream reader/writer.

**Describe the solution you'd like**
IPC stream reader/writer can read/write nested dictionaries in Map. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1582.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/ipc/reader.rs [hunk 1]
```diff
@@ -1481,4 +1481,55 @@ mod tests {
         let output_batch = roundtrip_ipc_stream(&input_batch);
         assert_eq!(input_batch, output_batch);
     }
+
+    #[test]
+    fn test_roundtrip_stream_nested_dict_dict_in_map() {
```
Comment (tustvold) [inline arrow/src/ipc/reader.rs:1486]:
```suggestion
    fn test_roundtrip_stream_nested_dict_of_map_of_dict() {
```

File: arrow/src/ipc/reader.rs [hunk 2]
```diff
@@ -1481,4 +1481,55 @@ mod tests {
...
+        let values = StringArray::from_iter_values(["a", "b", "c"]);
+        let keys = Int8Array::from_iter_values([0, 0, 1, 2, 0, 1]);
+        let dict_array = DictionaryArray::<Int8Type>::try_new(&keys, &values).unwrap();
+
+        let keys_array = Int32Array::from_iter_values([0, 0, 1, 2, 0, 1]);
+        let keys_field = Field::new("keys", DataType::Int32, false);
```
Comment (tustvold) [inline arrow/src/ipc/reader.rs:1492]: Might want to test dictionary encoded keys as well

File: arrow/src/ipc/reader.rs [hunk 3]
```diff
@@ -1481,4 +1481,55 @@ mod tests {
...
+        let values_field = Field::new_dict(
+            "values",
+            DataType::Dictionary(Box::new(DataType::Int8), Box::new(DataType::Utf8)),
+            false,
```
Comment (tustvold) [inline arrow/src/ipc/reader.rs:1496]: Could also test some NULLs

**Review by tustvold — APPROVED** [review]
This looks good to me, left some additional test suggestions

**Comment by viirya** [comment]
Thank you @tustvold. Revised based on the suggestions.

File: arrow/src/ipc/reader.rs [hunk 4]
```diff
@@ -1483,23 +1483,33 @@ mod tests {
...
     #[test]
-    fn test_roundtrip_stream_nested_dict_dict_in_map() {
-        let values = StringArray::from_iter_values(["a", "b", "c"]);
-        let keys = Int8Array::from_iter_values([0, 0, 1, 2, 0, 1]);
-        let dict_array = DictionaryArray::<Int8Type>::try_new(&keys, &values).unwrap();
-
-        let keys_array = Int32Array::from_iter_values([0, 0, 1, 2, 0, 1]);
-        let keys_field = Field::new("keys", DataType::Int32, false);
+    fn test_roundtrip_stream_nested_dict_of_map_of_dict() {
+        let values = StringArray::from(vec![Some("a"), None, Some("b"), Some("c")]);
```
Comment (tustvold) [inline arrow/src/ipc/reader.rs:1487]: Dictionaries store nulls in the keys and not the values, I'm not actually sure what this will do tbh

**Review by tustvold — APPROVED** [review]
Will get this in once CI finishes chugging along

**Comment by viirya** [comment]
Thank you @tustvold

**Pull request merged**